hbase-server/src/test/java/org/apache/hadoop/hbase/io/hfile/TestPrefetch.java

```diff
@@ -336,6 +337,62 @@ public void testPrefetchDoesntSkipRefs() throws Exception {
     });
   }
 
+  @Test
+  public void testOnConfigurationChange() {
+    PrefetchExecutorNotifier prefetchExecutorNotifier = new PrefetchExecutorNotifier(conf);
+    conf.setInt(PREFETCH_DELAY, 40000);
+    prefetchExecutorNotifier.onConfigurationChange(conf);
+    assertEquals(prefetchExecutorNotifier.getPrefetchDelay(), 40000);
+
+    // restore
+    conf.setInt(PREFETCH_DELAY, 30000);
+    prefetchExecutorNotifier.onConfigurationChange(conf);
+    assertEquals(prefetchExecutorNotifier.getPrefetchDelay(), 30000);
+
+    conf.setInt(PREFETCH_DELAY, 1000);
+    prefetchExecutorNotifier.onConfigurationChange(conf);
+  }
+
+  @Test
+  public void testPrefetchWithDelay() throws Exception {
+    // Configure custom delay
+    PrefetchExecutorNotifier prefetchExecutorNotifier = new PrefetchExecutorNotifier(conf);
+    conf.setInt(PREFETCH_DELAY, 25000);
+    prefetchExecutorNotifier.onConfigurationChange(conf);
+
+    HFileContext context = new HFileContextBuilder().withCompression(Compression.Algorithm.GZ)
+      .withBlockSize(DATA_BLOCK_SIZE).build();
+    Path storeFile = writeStoreFile("TestPrefetchWithDelay", context);
+
+    HFile.Reader reader = HFile.createReader(fs, storeFile, cacheConf, true, conf);
+    long startTime = System.currentTimeMillis();
+
+    // Wait for 20 seconds, no thread should start prefetch
+    Thread.sleep(20000);
+    assertFalse("Prefetch threads should not be running at this point", reader.prefetchStarted());
+    while (!reader.prefetchStarted()) {
+      assertTrue("Prefetch delay has not been expired yet",
+        getElapsedTime(startTime) < PrefetchExecutor.getPrefetchDelay());
+    }
+
+    // Prefech threads started working but not completed yet
+    assertFalse(reader.prefetchComplete());
+
+    // In prefetch executor, we further compute passed in delay using variation and a random
+    // multiplier to get 'effective delay'. Hence, in the test, for delay of 25000 milli-secs
+    // check that prefetch is started after 20000 milli-sec and prefetch started after that.
+    // However, prefetch should not start after configured delay.
+    if (reader.prefetchStarted()) {
+      LOG.info("elapsed time {}, Delay {}", getElapsedTime(startTime),
+        PrefetchExecutor.getPrefetchDelay());
+      assertTrue("Prefetch should start post configured delay",
+        getElapsedTime(startTime) <= PrefetchExecutor.getPrefetchDelay());
+    }
```

Comment:
This is really confusing. Why prefetch is supposed to start before the passed delay?

Reply:
In PrefetchExecutor class, passed in dealy is modified due to further computation. 

     if (prefetchDelayMillis > 0) {
        delay = (long) ((prefetchDelayMillis * (1.0f - (prefetchDelayVariation / 2)))
          + (prefetchDelayMillis * (prefetchDelayVariation / 2)
            * ThreadLocalRandom.current().nextFloat()));

Prefetch delay variation and a random seed is used for this computation. This reduces the "effective delay" compared to passed in delay. The same is passed to schedule(). 

(Due to this, in one of the earlier patch sets, I had introduced new class variable to be able to compare against "effective delay")

Hence, in the test, 

1. set the delay to 25000
2. ensurre that prefetch is not started till 20000 
3. In the window where prefetch is started and it's not complete, test that elapsed time is less than passed in prefetch delay

Reply:
Then just set the hbase.hfile.prefetch.delay.variation to 0 on your test. This way, you will have a deterministic delay of 25s.

Reply:
I have already tried that. If the hbase.hfile.prefetch.delay.variation is set to 0 in the test, it does not take effect.

Reply:
Per the formula below, if you set it to 0, it should pick exactly the value of `hbase.hfile.prefetch.delay`. So either your test is not setting the config properly, or the dynamic config logic isn't working properly.

`        delay = (long) ((prefetchDelayMillis * (1.0f - (prefetchDelayVariation / 2)))
          + (prefetchDelayMillis * (prefetchDelayVariation / 2)
            * ThreadLocalRandom.current().nextFloat()));`

Reply:
I agree that setting the variation to 0 would remove the random seed and delay would be deterministic. However, the cause for value of variation is not getting set is seems to be different instances of conf in the test and in the static initializer block of prefetch executor. 

 static {
    // Consider doing this on demand with a configuration passed in rather
    // than in a static initializer.
    Configuration conf = HBaseConfiguration.create();
  
With new changes if the prefetch delay is set, it's value get reflected becuase conf is getting reloaded which is not the case with the variation.

I am contemplating to code below to address this. Please let me know if you agree with the approach.

1. Add new test only method to set the prefetchDelayVariation in the PrefetchExecutor class 
2. Call it from the test to set to 0 and to unset 
3. With this prefetchDelayVariation cannot be declared as final in the prefetch executor class
4. Also, I am starting the timer abit late i.e. 
    HFile.Reader reader = HFile.createReader(fs, storeFile, cacheConf, true, conf);
    long startTime = System.currentTimeMillis();
    hence wait for small amount of time, say 500 millisecs before testing the assertions.

Reply:
why not also make it dynamic? Couldn't you just also set it in the PrefetchExecutor.loadConfiguration?

Reply:
For our initial requirement, making prefech delay alone dynamic is sufficient. 

prefetchDelayVariation can be updated in the loadConfiguration but it will be useful for testing only, atleast for now.

Reply:
I think, this is needed in certain cases where the user is expecting the exact delay before prefetch starts. With the delay variation, this delay is not fixed anymore which means the prefetch can trigger earlier or later depending on the calculated variance. Hence, IMO, this is not only a test only change, but it has its own merit. My 2 cents.

Reply:
I agree but this property is not externalised to the user, probably can be provided as a workaround in such cases. As mentioned above, I am trying out the change.